jenkins-master/jobs/scripts/workspace/runtests.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env python
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import glob
+import optparse
+import os
+import subprocess
+import sys
+
+
+here = os.path.dirname(os.path.abspath(__file__))
+
+
+class Runner(object):
+
+    def __init__(self, venv_path):
+        self.venv_path = venv_path
+        self.build_path = os.path.join(here, 'build')
+
+        os.chdir(here)
+        self.activate_venv()
+
+    def activate_venv(self):
+        if not os.path.exists(self.venv_path):
+            os.chdir(os.path.join(here, 'firefox-ui-tests'))
+            subprocess.check_call(['python', 'create_venv.py',
+                                   '--with-optional-packages',
+                                   self.venv_path
+                                   ])
+            os.chdir(here)
+
+        dir = 'Scripts' if sys.platform == 'win32' else 'bin'
+        env_activate_file = os.path.join(self.venv_path, dir, 'activate_this.py')
+
+        # Activate the environment and set the VIRTUAL_ENV os variable
+        execfile(env_activate_file, dict(__file__=env_activate_file))
+        os.environ['VIRTUAL_ENV'] = self.venv_path
+
+    def download_build(self, options, args):
+        command = ['mozdownload',
+                   '--destination=%s' % self.build_path,
+                   '--type=%s' % options.build_type,
+                   '--branch=%s' % options.branch,
+                   '--build-id=%s' % options.build_id,
+                   '--locale=%s' % options.build_locale,
+                   '--platform=%s' % options.platform,
+                   ]
+        subprocess.check_call(command)
+
+    def run_tests(self, options, args):
+        # Download the build under test
+        self.download_build(options, args)
+
+        import firefox_ui_tests
+        import firefox_puppeteer
+
+        command = [
+            'firefox-ui-update' if options.type == 'update' else 'firefox-ui-tests',
```

Comment:
As I said it looks ugly. I could create separate classes for them in that file. But when you wrote this I was thinking more about all that code and wonder if we are better to kill it all. We could make the run_tests.py script a simply activation script for the environment and let it execute a passed in command and all its arguments. That would make it way simpler, and we wouldn't need any other abstraction around.